### ISSUE TITLE
Create initial delimited import template

### ISSUE THREAD

**Issue by mcritchlow** [ucsdlib/horton#46]
### Descriptive summary

Create an initial csv import template to be used for initial tool R&D.  

Focusing on:
1. hierarchical components
2. linked records
3. CSV vs TSV vs Excel - is there a preferred format that will make life easier for DOMM and Dev?

Deliver:
* An initial template
* An initial set of unit and acceptance tests

### Dependency
#111 

### Related work

Ticket migrated from https://github.com/ucsdlib/dams5-cc-pilot/issues/39

**Comment by ghost** [comment]
@mcritchlow for "2. linked records", do you mean including URIs for concepts, like names and subjects?

**Comment by mcritchlow** [comment]
@GregReser yeah, exactly.

**Comment by ghost** [comment]
Would URI's be required?  We would need a pre-ingest tool to generate ARKs for all concepts already in the DAMS, then we would find URI's for the rest?

**Comment by ghost** [comment]
Maybe the fastest way to completing this initial import tool would be to duplicate the functions of the current tool, with a few improvements.  For URIs, we could do something similar to our pre-OLR ingest subject import, but expand it to agents.

**Comment by mcritchlow** [comment]
So that's a good question, and I think maybe gets at how the batch import tool fits into the overall DOMM workflow where you're presumably using something like Open Refine first. I'm not sure we need a separate tool. I think, similar to what we do now (though @lsitu would know better), we could do something like:

```
if cellValue is a URI
  try to lookup/match to an existing URI in the system
  if it exists, link to the existing URI
  if it doesn't exist. create a new local URI
if cellValue is a String
  try to find a URI using CV list (LCSH, Getty, etc.)
  if one is found, use it
  if not, create a new URI
```

Another variant on the String path is to just directly create a URI instead of doing remote lookup. This would be faster, of course. It would largely depend on whether you all would have done this ahead of time in OpenRefine or a similar tool. If you had, having our ingest tool do the lookup would be redundant and unnecessarily slow things down.

**Comment by mcritchlow** [comment]
@GregReser - I think that's a good practical approach.

Just noting here that if we do choose to use Excel as the file format, instead of CSV/TSV, I checked with Alex and for a different tool he used the following gems were useful:

* https://github.com/weshatheleopard/rubyXL (we could just use this if we can stick to `.xslx`
* https://github.com/zdavatz/spreadsheet (we would need this if we have to support `.xls`

**Comment by lsitu** [comment]
Thanks @mcritchlow for the information. 
@GregReser I think starting from something similar to our current support sounds good. Just FYI, here is a simple CSV example with just several fields that I set up for the purpose of testing the potential CSV batch ingest support in DAMS5: https://github.com/ucsdlib/horton/blob/develop/spec/fixtures/template_csv_import.csv

**Comment by remerjohnson** [comment]
The current workflow for OpenRefine reconciliation I have on a per-collection basis is if I have more than about 30 subjects, it is faster to just reconcile those in OpenRefine (otherwise there's manual lookups, recording the URIs, putting them in our Google Sheet post-ingest, etc.). And of course the larger the amount of subjects, the saved time goes up dramatically.

**Comment by arwenhutt** [comment]
Just another vote for sticking to current functionality (by and large) and leave external reconciliation tools for later.  

One of the things my review of the subject/name files is reminding me of is how unreliable automated reconciliation can be.  Even when there's an exact or strong string match, it can be an incorrect match, especially for names.  I think for now erring on the side of more manual and less automated, is prudent.

**Comment by mcritchlow** [comment]
I agree w/ @arwenhutt, and it sounds like we all do as well.

Given the conversations this morning on slack, do we want to, for now, also say we'll also be sticking with Excel `.xlsx` as the **only** supported file type for ingest? At least, for the MVP release of the batch import tool?

It sounds like the most directly useful file format to initially support, given the current DOMM workflow?

**Comment by ghost** [comment]
Yes, for now we should stick with the current functionality for ingest: string match against existing DAMS ARKs.  I would advocate for updating the subject import tool soon though so we can increase the speed and accuracy of that process.  We will also want to expand it to import Agents.

I'm OK with sticking with .xlsx for now.

**Comment by mcritchlow** [comment]
@GregReser - updating the existing subject import tool in DAMS4 or are you referring to the need for one in the new system? I'm trying to sort out what to do with that re: new ticket(s)

**Comment by ghost** [comment]
@mcritchlow Yes, in the new system.  I'd like to suggest a couple of improvement that might not require a lot of dev time.

**Comment by mcritchlow** [comment]
@GregReser Ok, so that's probably a separate ticket in this repo then for re-creating that feature. I'll get that setup.

In the context of this ticket, it sounds like we can proceed assuming the tool will be available to fill the role it currently does in the ingest process?

**Comment by remerjohnson** [comment]
I find supporting .xslx only suboptimal but it seems I'm being outvoted, so we can proceed on that.

**Comment by ghost** [comment]
@mcritchlow Yes, I can proceed with this ticket.

**Comment by ghost** [comment]
For Dates, the range is edm:TimeSpan and we using 3 properties label, begin, end.  I think we want to have these correctly associated with each other if there are multiple date predicates.  

We could have three columns for each date type:
Date:created Date:createdBegin Date:createdEnd

I'm thinking to minimize the number of columns by placing the three property values in a single cell:
label (begin | end)
Example:
Date:created = "between September 21, 2014 and April 5, 2017 (2014-09-21 | 2017-04-04)"

Any preference from the DEV or DOMM workflow perspective?

**Comment by lsitu** [comment]
@GregReser Are we still going to use the pie ( | ) for multiple values delimiting? If we are going to follow the current convention in dams4, then we've done something similar for Related Resource as cell delimiting with @ for multiple properties (description @ uri).

**Comment by ghost** [comment]
@lsitu I need to confer with others before I say how we will delimit multiple properties in the template.  I willget you an answer as soon as possible.

I will say that, because the the new template will be created and reviewed by people before ingest, we will want it to be human-readable, like the current ingest template.  This means some cell values will be user-friendly labels that are mapped to a URI in ingest or trigger the creation additional properties.
Some of the columns that will function this way are:
File Use
Type of Resource
Language

There will be new columns with this function, but I need a little time to write those out.

**Comment by remerjohnson** [comment]
I think what the issue is that pipe `|` is a "global" trigger to make things multiple values in the sheet. 

Therefore, we could do your thing above but instead use an `@` :

```
between September 21, 2014 and April 5, 2017 @ 2014-09-21 2017-04-04
```

But of course you could adjust the syntax as needed for convenience/clarity

**Comment by ghost** [comment]
@lsitu We will place multiple properties in a single cell as formatted text.  For now, only a couple of columns will use this, but we should plan for the future in case we use in in more columns.  It would be best to use the same cell format, so metadata preparation, review, and code development, are easier.

To make the cell text visually easier to understand, it would be good to clearly separate the label from the other properties.  It might also help to use name:value pairs for the properties.  This might help parsing on your end (you don't have to rely on the position of the value in an array, you can use the key value pair).  One concern is that a label might contain the character(s) used to indicate properties.  Here is my proposal:

label @{name:value | name:value}

date:creation
between January 1 - August 3, 1925 @{begin:1925-01-01 | end:1925-08-03}

related resource:depiction
1925 Trip - Album @{uri:http://library.ucsd.edu/dc/object/bb35511535}

Questions:
1. Is "@{" a good way to uniquely indicate that the text afterwards are key value pairs?
2. I used " | " to separate name:value pairs hoping it will be unique (won't appear in the label)

Do you see any problems from the coding point of view?
@arwenhutt @remerjohnson  any thoughts on our workflow?

**Comment by remerjohnson** [comment]
@GregReser I do think, as I and Longshou noted above, that we should keep the pipe as a "global" multi-value indicator, while @ is more of a per-cell rule, therefore, we'd have to find some other way to delimit the begin/end date values within the value. I'm sure @lsitu might have some ideas on that.

**Comment by ghost** [comment]
I was thinking the name:value pairs were repeated within the { } so pipe made sense.  That's a pretty loose definition though :)
A comma is too common to work.  I don't think we want to place the value in quotes, e.g., begin:"1925-01-01", because we don't use quotes anywhere else and it might cause confusion (When do I have to use quotes?)

Is " , "  safe?  
Other ideas?

**Comment by remerjohnson** [comment]
@lsitu can correct me if I'm wrong, but since we are calling out key-value pairs, `begin:` and `end:` seem to be fine to indicate a value is bound to those keys. But again I may be wrong there. So:

```
date:creation
between January 1 - August 3, 1925 @{begin:1925-01-01 end:1925-08-03}
```

**Comment by lsitu** [comment]
@remerjohnson @GregReser It does sound good to me for supporting a simple case like that. 
But if we are going to use name/value pairs used in cell properties after @ with a more general support, maybe we can balance a machine readable format like JSON ({key1:"value1", key2:"value2", ...} ), Hash ({:key1=>'value1', :key2=>'value2', ...}) etc.

**Comment by ghost** [comment]
@lsitu I'm concerned that placing the value in quotes might cause confusion because we don't use quotes around values anywhere else.  The questions would always arise: "When do I have to use quotes?"

**Comment by remerjohnson** [comment]
I _think_ that the context of it being within the curly brackets `{}` would allow that to work (i.e., those brackets set up basically JSON syntax rules. 

But agree if that isn't the case, I wouldn't want to always have to escape double quotes like `\"\"`

**Comment by lsitu** [comment]
Yes, I agree. It's just related to how we interpret the values within the curly brackets {}. It will be nice to use a pattern that is more human readable and machine readable.

**Comment by ghost** [comment]
Yes, it's inspired by JSON, but made more human readable and following some of the conventions of our template.

I would like to avoid using quotes because it adds another rule and some clutter for humans.
@lsitu would it be possible to parse  this, as Ryan suggested?
@{begin:1925-01-01 end:1925-08-03}

In this case, the keys are the separators

**Comment by lsitu** [comment]
@GregReser Yes, it'll work for the case like @{begin:1925-01-01 end:1925-08-03}. But we may need to know about the keys in advance when we are making it "the keys are the separators". Do you have other use cases for the using of key/value pairs in the curly brackets {}? And do you have any thoughts regarding extending the use of this key/value pairs encoding in the future?

**Comment by ghost** [comment]
I think we should apply this to related resource to be consistent.  That will probably be all right now.

Locus 920, Area F, Area Stratum BR, BEDROCK CELL 8 @{uri:http://library.ucsd.edu/ark:/20775/bb7987903k}

We should have a controlled set of keys and these should part of the pre-ingest validation.  If an unrecognized key within {} is detected, an error should be reported.  I will supply this list of keys with the new template.

**Comment by mcritchlow** [comment]
Feel free to add that requirement (controlled key list) in the description of this ticket as you see fit @GregReser, I think it would be good to have detailed there.

**Comment by ghost** [comment]
Last (I hope) question before I pass the first draft of the new template for dev...

Do we have a need for more than 2 files per object/component/sub-component?
Should we fold file and file use into a single column so we can repeat it as many times as necessary?

myFile.tif @{fileUse:image-source}

**Comment by lsitu** [comment]
@GregReser It sounds good to me. What's the rules to identify all the keys within {}? 
When moving forward with the fileUse properties, it looks like we may need to map dams4 fileUse properties to PCDM file use (http://pcdm.org/2015/05/12/use) and add additional file use properties for DAMS5. @mcritchlow Do you have any suggestions on it?

**Comment by mcritchlow** [comment]
No specific suggestions other than agreeing that if we don't already have a mapping for those properties we should do so. Probably in the data dictionary? I think @remerjohnson may have started looking into this?

**Comment by arwenhutt** [comment]
@GregReser Do we have a need for more than two files per component? 

My inclination is to leave values in separate columns as much as possible, unless there's a compelling need for a more complex encoding.

**Comment by remerjohnson** [comment]
@mcritchlow I have, but a complete mapping is not possible now since we don't just have the idea of "use" in relation to objects, but also specific rules about what happens to certain things like documents/audio/video. So if a rough mapping of say `document-service -> pcdmuse:ServiceFile` and `data-service -> pcdmuse:ServiceFile` is fine (note the many cases of overlap and lossiness), I could do that mapping. 

A missing piece here is PCDM was treating a thing we conflate as separate things, and were going to work on "[Object types and Profiles](https://github.com/duraspace/pcdm/wiki/Examples-&-Profiles)", which apparently is still a work in progress.

**Comment by ghost** [comment]
I've had one collection that wanted more than two files per component, but we ended up zipping the files into one, so it was not a problem.

Agreed - we'll stick with the current separate columns for file and file use.

**Comment by ghost** [comment]
@lsitu In the dams4.2 data dictionary Google sheet, we had specified a way to format identifiers.  Do you recall if we are going to use that or were there discussions about modifying it?
Example: urn:x-public-id:doi:10.1000/182

For the ingest spreadsheet, there will be a column for each identifier type, e.g., identifier:doi and the cell value will be just the identifier value, e.g., 10.1000/182.  On ingest, the complete formatted string will be constructed.  So my question is about how to specify what the final ingested value should be.

[https://docs.google.com/spreadsheets/d/1OjTm1Kuzo-An-THdpUkjwa5puul-b4Lx3SK0lB6rvl8/edit?usp=sharing](url)

**Comment by lsitu** [comment]
@GregReser I don't recall that we have any further plans to discuss and modifying it so I assume that we are going to use it. And as your example showing, the values in the "URN Example" column should be the format of the final ingested value of the identifiers. @mcritchlow Could you correct me if it's wrong?

**Comment by mcritchlow** [comment]
@lsitu Yep, I think that's correct as far as the current state of that discussion.

I admittedly wonder if it's ultimately easier to just have separate properties for each identifier type, for dealing with things like indexing, display order, access control, etc. But I can also see the value in just moving forward with what's already documented.

**Comment by remerjohnson** [comment]
@mcritchlow Yeah we had a little back-and-forth on that. Like listing bare ORCIDs in its own field seems... disjointed unless those could get labels pulled

**Comment by mcritchlow** [comment]
@remerjohnson What do you mean by labels pulled? The current spreadsheet seems to be using a static/fixed label for each identifier type. So every ORCID will be displayed as:

ORCID http://orcid.org/0000-0003-3270-1306

Or something. The ORCID label could be whatever we wanted, but it would be the same display label for every ORCID instance, right?

**Comment by remerjohnson** [comment]
@mcritchlow yeah going to think about this more. ORCIDs should obviously be binded to Agents/authorities, so going to think about how that would happen in the template

**Comment by mcritchlow** [comment]
So @remerjohnson & I had a quick Slack convo about this. I think ORCID's are a bit of a special case here compared to other identifier types, and I wonder if they should be bracketed off.

Primarily because ORCID's are not associated with Objects/Works, they are associated with People/Organizations (Agents). So I think the assertion that Ryan made that it's disjoint is correct. Because it seems odd to model an Object as having three creators and then three ORCIDs. More accurate would seem to be an object that links to three researchers (Agents) that themselves have ORCIDs.

**Comment by ghost** [comment]
It does seem like we should place ORCID in the Agent record.  Can we create a way to display this in the new UI?  It could display beside the agent name, as a hover, as an "ORCID" button.  It could also be part of a more comprehensive "info" pop-up that displays other things, like a Wikipedia snippet. 

If we could add this to the new UI from the beginning, it would solve a couple of problems:
1. Clearly associate ORCID with Agent label
2. Provide the visible benefit of linked data we've been asking for.  We would be a shining example of linked data in real world use.

**Comment by mcritchlow** [comment]
@GregReser Yeah, I mean I defer to @lsitu, but I think we're talking about indexing and display rules at that point. Similar to how we display components in the object view.

**Comment by lsitu** [comment]
@GregReser I agree with Matt to talk about the UI issues later at the point with indexing and display rules. But are we going to add the ORCID to the Object or the Agent? If in the agent, which field for ORCID?

**Comment by ghost** [comment]
I just bring up implementing the ORCID display in the UI sooner rather than later so it's not hidden metadata.  Kind of planting the idea as a priority.

ORCID will go in the Agent authority record as closeMatch.

**Comment by mcritchlow** [comment]
Is there a predicate/namespace for ORCID we could use instead of closeMatch? If something like that already exists, might be worth using and more explicit? It might also make targeting ORCID vs other properties in the Agent class a bit easier for specific display/indexing rules. But @lsitu would know better if that's an actual concern or not.

**Comment by ghost** [comment]
@mcritchlow funny you say that, we were just talking about using exactMatch when we are very confident.

For targeting ORCID vs other properties in the display, can we use the base URL, e.g., http://orcid.org ?

**Comment by lsitu** [comment]
@mcritchlow Yes, I am concerning about it either. I think it'll be nice if the ORCID can have its own predicate so that we can grab it for special handling.

**Comment by remerjohnson** [comment]
@lsitu @mcritchlow the makers a fabio have an ontology called "scoro": http://www.sparontologies.net/ontologies/scoro 

One of the predicates is `scoro:hasORCID`

We could put our property within Agent as `hasORCID` or `identifier:orcid` and point to that predicate?

**Comment by ghost** [comment]
@lsitu I have some details in the new template to work out, but a draft version for discussion is available at:
https://docs.google.com/a/ucsd.edu/spreadsheets/d/1EhMWaDFsBZDhp-BU-5LtOI0tECDx2EtAZ28iZk-Ih58/edit?usp=sharing

Cells in yellow are to be determined, so you can ignore them for now.
The header list is vertical so I can show you the definitions and sample values.  The final result will be a horizontal headers like the old ingest spreadsheet.

The tab "CV values" contains values allowed in controlled value columns.  As in the old ingest spreadsheet, the values are labels that should be converted upon ingest.  For example: CC-BY  = http://creativecommons.org/licenses/by/4.0/
I have added the matching values for convenience, but you can also find them on the "dams4.2 data dictionary" Google sheet.

The tab "Formatted key-value rules" are the values allowed within { } for single-cell multi-property strings.  I don;t think we need individual validation for each  { } column (date, relatedResource).  I expect that the keys will be applicable any time we use the { } format.

There will be more agent:role columns added.  I have to get these from our other documentation.

Let me know if you have questions.  If we have agreed on most things, I can turn this into a .xlsx template for your development testing.

**Comment by lsitu** [comment]
@GregReser The ingest template looks nice. Thank you very much.

- For CVs controls, is it possible to use exact CV without conversions or not?
- For key/value inside {}, Is there a generic way to identify all the key/values pairs or not? For example, a comma, a while space etc. I see you mention the "Formatted key-value rules" and it could be any keys, right?
- For linking of the range ucsd:Agent, what's the look up rules for agent types like Agent:Person, Agent:Group and Agent:Organization, just using the label and ignore all other fields? 
- In "Ingest Instructions" for IDs, it looks like the value "convert to:  urn:x-public-id:accession:[cell value]" is a little misleading since prefix "urn:x-public-id:accession" could be vary for each identifier type. Maybe using something like "urn:[x-experimental-id]:[URN Prefix]:[cell value]" is better?

**Comment by ghost** [comment]
@lsitu 
- CVs can only use the short label not the conversion
- For key/value inside {} - at this time, the only valid keys are the ones listed in "Formatted key-value rules".  key/values pairs are separated by white space.  Would it work to consider the white space+key as a reliable separator?  Thinking ahead, if we have a white space in a value, will the parsing still work?
- For linking of the range ucsd:Agent - lookup label and alternateLabel only.  Is that what you mean?
- I changed the "Ingest Instructions" for ID the correct prefix to indicate public or private and I added the URN Prefix.  Is that useful?

**Comment by lsitu** [comment]
@GregReser :
- CVs can only use the short label not the conversion
*I mean values in CVs for validation should be matched what are allowed to be used in the Excel spreadsheet for ingest. Another way that we can do is to retrieve the CVs from the local authority repository. But we may need to define those CVS in label/url pairs and use either the label or the url in the Excel spreadsheet. What do you think?
- For key/value inside {} - at this time, the only valid keys are the ones listed in "Formatted key-value rules". key/values pairs are separated by white space. Would it work to consider the white space+key as a reliable separator? Thinking ahead, if we have a white space in a value, will the parsing still work?
*Hmm, lots of concerns. I am worrying about the "description" text that could be anything, and the space+key could be anything that need to validate, right?
- For linking of the range ucsd:Agent - lookup label and alternateLabel only. Is that what you mean?
*Yes. I think this is just something that we may need to know for questions like:
 What is a duplicate record? How those alternateLabel and other fields will be used? 
 Is it possible to have the same label for different type of agents or not?
- I changed the "Ingest Instructions" for ID the correct prefix to indicate public or private and I added the URN Prefix. Is that useful?
*Yes, it's clear now and I think it will help others understand it. Thanks.

**Comment by lsitu** [comment]
@GregReser @remerjohnson For key/value inside {} , maybe we can just define a separator char, something like comma, semicolon etc. that are used less often. If we need to use it in side a "description" text, we can escape the char with \ just like what we are currently doing with pie ( | ). Once we can determine the key/value in pairs, then we can just split them for the key and value. What do you think?

**Comment by ghost** [comment]
@lsitu 
- CVs
For simplicity and readability, only the label should be valid in the Excel spreadsheet.  You have a method now that says the spreadsheet label for language "eng - English" is valid.  On ingest the value "eng" is used in the object build.  This is the same idea.
- For key/value inside {} - maybe space+key is not the right way to talk about it.  Basically, we have a quasi JSON structure, except we have no quotes around the value and no comma between the key:value pairs.  This makes it challenging to parse the contents of {}.  Before we say that we have to use a JSON structure, let's see if we can make the simpler form work.
  - Our keys are limited.  Right now there are only three valid keys - begin: end: uri:  Provided these key substrings are not found in any values within {}, can the keys serve as separators?  For example, if the substring "end:" is found, then the substring that follows (until "}" or another valid key) is the value associated with end:  As I write this, it seems complicated and may not work.  The benefit would be the parsing could be used in future multi property cells.
  - An alternative would be to have parsing rules for each multiple properties column.  We know what to expect in each column.  Right now we have two columns like this:
    - related resource we already know because we are using it in the old spreadsheet.  Just modify the parsing for the use of "@{uri:value}.  I mean the value = the substring after "{uri:" and before "}"
    - date will always have the same format @{begin:DateTime end:DateTime} so the separator will always be " end:"
   - Another alternative would be to use JSON.  My fear there is that the quotes around values might cause trouble with how quotes are handled elsewhere.  For instance, if we use CVS files for ingest.  Is this a concern?

**Comment by lsitu** [comment]
@GregReser Just a couple of little questions regarding the two issues above:
For CVs, maybe we just need the instructions to build the values for each field and ingest it into the Fedora. For example, with "eng", how would you like the uri to be?

For key/value inside {} , don't you expect our program to report any wrong keys? I think we can just make it work anyway. But with the simple key/value cases that you have now, if we can say white space is the separator for multiple key/value pairs, then it'll become much more general, thoughts?

**Comment by ghost** [comment]
@lsitu if we need a reliable separator that could be used for future cases, then I want to use " | ".  Whitespace is too generic and common in values.  I know we use " | " to indicate multiple values in other cells, but I think the use inside {} is similar: we are separating multiple key:value pairs.

We could use another delimiter, e.g., "; " within {}, but anything we think of might be contained in another string somewhere in the spreadsheet.  At least with " | " we are aware of it's use as a separator.

**Comment by ghost** [comment]
@lsitu For CV values build values
- I have added the full URIs for language.
- copyright jurisdiction uses a literal string and will be a two character code as shown.

Are there any others that need clarification?

**Comment by ghost** [comment]
@lsitu For linking of the range ucsd:Agent - lookup label and alternateLabel

Your questions are very good and difficult to answer.  Here's my first thoughts:
- What is a duplicate record?
  - By running the new authority import with pre-ingest validation (issue 48) we should be able to resolve duplicate records.  Duplicates might still slip through though.
  - For the Object ingest Agent matching (this issue)
    - Can the pre-ingest validation report multiple string matches?  If so, we could resolve them before final ingest
    - On final ingest, the last match could be used, but the report could indicate which other matches were found.
- How those alternateLabel and other fields will be used?
   - They will be used for search ("Central Library" is the same as "Geisel Library"), possibly, maybe, hopefully
   - They will be used for display and web linking in the UI, possibly, maybe, hopefully
- Is it possible to have the same label for different type of agents or not?
  - If we are going to rely on string matches, labels will have to be unique.  We will have to manage this by pre-ingest testing and through periodic authority record analysis (run a report to search for duplicates).

**Comment by lsitu** [comment]
@GregReser Thanks for clarifying all these. 
For linking of the range ucsd:Agent - lookup label and alternateLabel
*****If we lookup records with label and alternateLabel, then we need both the label and alternateLabel in the spreadsheet, and we'll allow different combination of label and alternateLabel of ucsd:Agent records to be created. This means, the label won't be unique, which will be conflicting with your assumption for the last question below.

- What is a duplicate record?
  - By running the new authority import with pre-ingest validation (issue 48) we should be able to resolve duplicate records. Duplicates might still slip through though.
     *****I think determine a matched record (or duplicate records) is important as you know we'll exclude duplicate records from created. And if we found any duplicate records for some reason, we need to fix our tool as a bug and clean up the metadata at the meantime.
  - For the Object ingest Agent matching (this issue)
    - Can the pre-ingest validation report multiple string matches? If so, we could resolve them before final ingest
        *****Yes, we can report multiple string matches in the pre-ingest validation.
    - On final ingest, the last match could be used, but the report could indicate which other matches were found.
        *****I think we can report it in the ingest log.

- How those alternateLabel and other fields will be used?
  - They will be used for search ("Central Library" is the same as "Geisel Library"), possibly, maybe, hopefully
  - They will be used for display and web linking in the UI, possibly, maybe, hopefully
- Is it possible to have the same label for different type of agents or not?
  - If we are going to rely on string matches, labels will have to be unique. We will have to manage this by pre-ingest testing and through periodic authority record analysis (run a report to search for duplicates).
     *****If we lookup records with label and alternateLabel, then label won't be unique.

**Comment by lsitu** [comment]
@GregReser For key/value pairs in {}, I think we cannot use pie ( | ) as the separator since we've used it to delimit multiple values, unless we can clarify either key/value pairs or multiple values will be allow in a cell, but not both. 
It seems like "; " is a better fit than the space. If anyone don't like it, let's move forward with the origin plan and we'll make it work for now then. Thanks.

**Comment by ghost** [comment]
@lsitu For key/value pairs in {}, you may be right, but if we use "; " as a separator within {} won't that cause a problem when we have "; " as part of a string in other cells?  In other words, we would treat "; " as a separator only some times.

**Comment by lsitu** [comment]
@GregReser No, the separator for key/value pairs in {} only affects the values inside the {}. It'll be safe to use it anywhere else in the spreadsheet. For example, if we choose to use ";" as the key/value pairs separator, we can still use it intact outside of the {} without escaping in the current cell or any other cells.

**Comment by lsitu** [comment]
For example, if we choose "," as separator, the following cell values are valid:
Jan. 1st, 1925
Jan. 1st, 1925@{begin:1925-01-01, end:1925-08-03}
Jan. 1st, 1925@{begin:1925-01-01, end:1925-08-03} | Jan. 1st, 1925

**Comment by ghost** [comment]
@lsitu I assume you are thinking that a multi property cell could contain more than one value?  You want to be able to split the multiples first then parse the key:values inside {}
Example date:issued
April 8, 1908 @{begin:1908; end:1908} | June 30, 1935 @{begin:1935; end:1935}

OK, let's use "; "  Since we are changing this, how about using key=value for readability?  :  and ; look very similar.
I'm taking inspiration for the Dublin Core formatted text string spec: http://dublincore.org/documents/dcmi-dcsv/   Our use is a little different.

Example:
April 8, 1908 @{begin=1908; end=1908}

**Comment by ghost** [comment]
@lsitu I missed your example above showing " | " when I was writing my comment.
I had not planned on allowing multiple values in a {} cell, which is why I didn't see a problem.  I guess we should allow multiple values, since that is our general rule.  So yes, let's use "; " inside {}

**Comment by lsitu** [comment]
@GregReser Since we are not following the JSON format for key/value pairs in {}, it sounds good to use ";" as key/value pairs separator and and use "=" as key/value delimiter. I think your example looks good: 
April 8, 1908 @{begin=1908; end=1908}

**Comment by ghost** [comment]
@lsitu Great, I'll update the template.

**Comment by ghost** [comment]
@lsitu Matching records by string - Right, we don't want duplicate ARKs for the same Agent, Place, or Concept.  It seems that we want to lookup all labels and alternateLabels to make sure there is not a match or *possible match*.  If we only search labels, we might miss a match.  The possible matches will require that a person makes the final decision.

In the ingest spreadsheet, I don't think we can include both label and alternateLabel partly because we wouldn't know what the alternates are - we only know what the data provider gave us.  If they gave us "Reser, Greg" and there is an ARK with alternateLabel "Reser, Greg" and label "Reser, Gregory" we want the report to say: possible match found with label "Reser, Gregory".  These would be flagged for human review.  Would it possible to make something like this work?

**Comment by lsitu** [comment]
@GregReser Maybe we can do something with non-work records pre-validation in [Support batch ingest of non-Work records](https://github.com/ucsdlib/horton/issues/48) to identify all the records with the same label?
Once an authority record with a label and an alternateLabel is identified to be used, then we have to include both the label and alternateLabel in the ingest spreadsheet. If we only include the label in the ingest spreadsheet, we'll never know which record is used for ingest if there are several authority records that have the same label exists.

**Comment by ghost** [comment]
@lsitu Perhaps handle all the records with the same label or alternateLabel in the non-work  (authority)  records pre-validation and only match label in the object ingest.  This puts the burden on us to do the proper pre-ingest validation and conflict resolution, but it makes the final ingest easier.  Adding alternateLabels to the final ingest spreadsheet would be messy, so it would be better to avoid it.

On final ingest, we should get a report of all new ARKs generated (as we do now) and we should add the Class and label that was created.  These could be reviewed manually if anything looks wrong.

**Comment by lsitu** [comment]
@GregReser Yes, I agree. So in the final ingest, only the label will be used in the spread sheet we will lookup and use any authority record that matched the label in the spreadsheet. I think we can extend it to use the @{} syntax to include the known alternateLabel later when it's necessary. Does it sounds good?

**Comment by ghost** [comment]
@lsitu Sounds good!

**Comment by mcritchlow** [comment]
@GregReser @lsitu - I know we've shifted focus to other things lately, but I wanted to circle back to this effort and see if it would be possible to articulate what remaining gaps we have for this ticket, whether that work can be broken down into individual tickets, etc. Curious what you both think, thanks.

**Comment by ghost** [comment]
@mcritchlow I think these are the remaining gaps:
- resolve relatedResource predicate(s) and Class #58 
- resolve rights properties
  - do we need copyright status or can it be inferred from rights URI?
- resolve Matt's question about identifier:roger record (should it be more generic, e.g., "identifier:catalog_record"
- add remaining agent predicates (it's a long list and I was waiting because it will make the template harder to read).

@lsitu In the object_import_template, I've been highlighting changes in yellow to make sure they you see them.  I suggest you remove the yellow when you have implemented the change.  This way we will all know what remains to be updated.

Anything else?

**Comment by lsitu** [comment]
@GregReser It looks like there could be more issues. See my comments in the attached Excel spreadsheet in https://github.com/ucsdlib/horton/issues/69#issuecomment-291249356.
And also, we may need instructions for assembling the title column fields into a big title string, right?

**Comment by ghost** [comment]
@lsitu good question about title.  Are you thinking of how we currently assemble title, part name, part number, subtitle in mads:authoritativeLabel?  I don't think we are doing this in D5.

**Comment by lsitu** [comment]
Yes. When I first see those bunch of title fields. Then I think it'll be just the UI displaying issue that we need to talk about later. Thanks.

**Comment by ghost** [comment]
You're right, it will be a messy display issue :)

**Comment by lsitu** [comment]
@GregReser It looks like we can't use property name "visibility" (row #134), which is hold as keyword in Hyrax. The following error is thrown from Hyrax:
ArgumentError: visibility is a keyword and not an acceptable property name.
  /Users/lsitu/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/active-fedora-11.0.0.rc7/lib/active_fedora/attributes.rb:127:in `define_active_triple_accessor'
  /Users/lsitu/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/active-fedora-11.0.0.rc7/lib/active_fedora/attributes.rb:118:in `property'
...

Do you have any thoughts to avoid it?

**Comment by ghost** [comment]
@remerjohnson @arwenhutt Should we change visibility and visibility expiration to:
rights override
rights override expiration

**Comment by ghost** [comment]
@lsitu Let's go with:
rights override
rights override expiration

**Comment by lsitu** [comment]
Sure and it's done. Thanks Greg.

**Comment by gamontoya** [comment]
**Note:** Waiting on FileUse modeling

**Comment by gamontoya** [comment]
Defer for Batch/Import/Export Working Group.
